Recycle ServletChannel at ServletContextHandler completion.

gregw opened the pull request:
Fix bug with #10801 in that completion listeners were not well sequenced with stream wrappers.

gregw committed:
Recycle ServletChannel at ServletContextHandler completion.

Fix bug with #10801 in that completion listeners were not well sequenced with stream wrappers.

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12.0.x/onCompletionEvents

lachlan-roberts commented on jetty-core/jetty-server/src/main/java/module-info.java:
```diff
@@ -28,4 +28,5 @@
 
     exports org.eclipse.jetty.server.jmx to
          org.eclipse.jetty.jmx;
+    exports org.eclipse.jetty.server.internal;
```
You shouldn't be exporting the internal package.

If you really need to do this you should export it specifically to the module where you need it.


Do you even need this seems all usages are within `jetty-server`.

gregw replied:
Nope... dang intellij!

gregw replied:
I do find the whole jpms things strange.... previously I would have just made this class package private and it would have worked fine... but now I have to move it into a different package 'internal' and make it public so I can use it, but then use a module.info file to restrict who can use it.    Not sure I see the value in a public/restricted org.foo.internal.Bar vs package protected org.foo.Bar?

lachlan-roberts commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -345,50 +344,12 @@ default void push(MetaData.Request resource)
-                }
-            });
-        }
+            if (stream instanceof CompletionStreamWrapper completionStreamWrapper)
```
Maybe you could search through all the stream wrappers to see if one is a `CompletionStreamWrapper` and add it. Just in case there is a different wrapper added inbetween.

gregw replied:
@lachlan-roberts that would defeat the purpose.  If you skip through other wrappers, then you change the order of events.

gregw committed:
Recycle ServletChannel at ServletContextHandler completion.

Fix bug with #10801 in that completion listeners were not well sequenced with stream wrappers.

gregw committed:
Merge branch 'jetty-12.0.x' into fix/jetty-12.0.x/onCompletionEvents
